GameLogic/python_emulator.py

```diff
@@ -0,0 +1,531 @@
+        self.min_moving_speed = self.dino_width / (self.JUMP_TIME * 5)
+        self.real_moving_speed = int(
+            self.min_moving_speed * self.ADDITIONAL_MOVING_SPEED
+        )
+
+        # Set up display
+        self.screen = pygame.display.set_mode((self.WIDTH, self.HEIGHT))
+        pygame.display.set_caption("DinoJump!")
+        self.clock = pygame.time.Clock()
+
+        # Game Attributes
+        self.start_game = False
+        self.is_right_foot_down = True
+        self.is_jumping = True
+        self.jump_start_time = 0
+        self.is_ducking = False
+        self.cannot_spawn_object = True
+        self.cooldown_start_time = 0
+        self.cactus_locs = []
+        self.mid_bird_locs = []
+        self.high_bird_locs = []
+
+    #############################################################################
+    # Drawing Functions
+    #############################################################################
+
+    # Draws the black line for the ground
+    def draw_ground(self):
+        for pixel in range(self.WIDTH):
+            pygame.draw.rect(
+                self.screen,
+                self.BLOCK_COLOR,
+                (pixel, self.GND_AXIS, self.GRID_SIZE, self.GRID_SIZE),
+            )
+
+    # Takes whichever sprite is dino_array currently and it's location and draws to the board
+    def draw_dino(self):
+        for row in range(self.dino_height):
+            for col in range(self.dino_width):
+                if self.dino_array[row, col] == 1:
+                    pygame.draw.rect(
+                        self.screen,
+                        self.BLOCK_COLOR,
+                        (
+                            self.dino_x + col * self.GRID_SIZE,
+                            self.dino_y + row * self.GRID_SIZE,
+                            self.GRID_SIZE,
+                            self.GRID_SIZE,
+                        ),
+                    )
+
+    # Determines which dino is shown: running LR, jumping, ducking LR, and game over (TODO)
+    def determine_and_draw_dino(self):
+        self.curr_walk -= 1  # counting down
+        if self.is_ducking and self.curr_walk == 0:
+            self.curr_walk = self.UPDATE_WALK_FRAMES
+            if self.is_right_foot_down:
+                self.is_right_foot_down = False
+                self.dino_array = self.ducking_left_array
+            else:
+                self.is_right_foot_down = True
+                self.dino_array = self.ducking_right_array
+        elif self.is_jumping:
+            self.dino_array = self.jump_array
+        elif self.curr_walk == 0:
+            self.curr_walk = self.UPDATE_WALK_FRAMES
+            if self.is_right_foot_down:
+                self.is_right_foot_down = False
+                self.dino_array = self.running_left_array
+            else:
+                self.is_right_foot_down = True
+                self.dino_array = self.running_right_array
+        self.draw_dino()
```

Comment:
This block of code seems to be doing two things. It updates the "state" of the game (such updating `self.is_ducking`), and it changes the "display" of the game (such as changing `self.dino_array` and calling `draw_dino()`).
I think it might be worth considering to distinguish between state updates and display changes, by putting them into different functions, doing all the state updates first, then change the display based on the new state (or the other way around works too, there is a game loop so it doesn't really matter). The code can be made more readable to by keeping them modular.